### ISSUE TITLE
Support CSS in tagged template literals

### ISSUE THREAD

**Issue by philpl** [prettier/prettier#1948]
Since initial CSS support with PostCSS has landed (https://github.com/prettier/prettier/pull/1636), it'd be great to consider formatting CSS inside tagged template literals :)

A couple of projects could benefit from this, namely:

- https://github.com/threepointone/glam
- https://github.com/tkh44/emotion
- https://github.com/styled-components/styled-components
- more I'm forgetting?

(Opened after a brief discussion on Twitter: https://twitter.com/Vjeux/status/871108131278864384)

**Comment by vjeux** [comment]
Would be awesome indeed! And shouldn't be super hard to do.

**Comment by rauchg** [comment]
We would love this for `styled-jsx`:
https://github.com/zeit/styled-jsx

**Comment by vjeux** [comment]
If someone is motivated, that would be an awesome contribution!

**Comment by pomber** [comment]
I really want this feature! Do you think this can be done in one day? If so, I'll try to have something working between tomorrow and Monday.   
Any tips are welcome.

**Comment by azz** [comment]
I'd love to see this problem solved in a way that makes it easy to implement arbitrary parser switching within prettier itself, rather than a "hack" for just CSS.

Determining whether a template string is CSS or not may be determined in different ways. For styled-components we could probably use something like `escope` to determine it based on the tagged template function. Similar for other projects, but the criteria may vary wildly.

Vue.js and GraphQL are the other projects that comes to mind that could greatly benefit from supporting multiple languages in one run of prettier.

I'm thinking something along the lines of an AST visitor/interceptor that would run during `genericPrint()`. 

For maintenance purposes it would also be great to keep all this code together.

I was thinking something along these lines:

```js
function getSubtreeParser(path, currentParser) {
  const node = path.getValue();
  const parent = path.getParentNode();

  switch (node.type) {
    case "TaggedTemplateExpression":
      // styled-components
      if (
        node.tag.type === "MemberExpression" &&
        node.tag.object.name === "styled"
      ) {
        return "postcss";
      }
      // graphql-tag
      if (
        node.tag.type === "Identifier" &&
        node.tag.name === "gql"
      ) {
        return "graphql";
      }
    case "TemplateLiteral": {
      const parentParent = path.getParentNode(1);
      // styled-jsx
      if (
        node.quasis.length === 0 &&
        parent.type === "JSXExpressionContainer" &&
        parentParent.type === "JSXElement" &&
        parentParent.openingElement.attributes.some(
          attribute => attribute.name.name === "jsx"
        )
      ) {
        return "postcss";
      }
    }
    default:
      return currentParser;
  }
}
```

**Comment by giuseppeg** [comment]
Something along the lines of what @azz described would be interesting.
I would probably define *matchers* as partial visitors which traverse bottom up starting from a `TemplateLiteral`. I'd make an array of them and use `some` rather than go for a big `switch`. 

One thing to keep in mind is that template literals can have expressions therefore the string to format should be joined with a placeholder e.g. `quasis.join('__prettier_something__')` and the split once formatting is done. Also indendation should be taken into account.

@vjeux am I right that this should be done here: https://github.com/prettier/prettier/blob/master/src/printer.js#L1735 ?

**Comment by vjeux** [comment]
On this topic, we're also doing intermingled parsing within the postcss parser. https://github.com/prettier/prettier/blob/master/src/parser-postcss.js#L182-L184

postcss relies on postcss-[values/media-query/selector]-parser for subsets. Would be nice to have a system that handles all of them at once. It would be great for error location reporting and cursor placement :)

**Comment by josephfrazier** [comment]
There's a related idea in https://github.com/prettier/prettier/issues/1674#issuecomment-303794974:

> One way to fix this is to use a parser that parses both the HTML and the JS/CSS into a single AST. Then, `prettier --range-start 71 --range-end 167` should work. Eventually, if prettier supports formatting HTML, having a unified AST would allow it to format embedded JS and [CSS](https://github.com/prettier/prettier/pull/1636). I'm not sure if such a "mixed-mode parser" exists, but it might not be hard to implement as follows:
> * Parse HTML into an AST ([example](https://astexplorer.net/#/gist/e0a8931806d3f1d97ed65516c7e2f5d1/b656133dd1e0290d28b521e14e38ecc824a36126))
> * Traverse the AST, replacing the children of `script` nodes with the corresponding JS AST ([example](https://astexplorer.net/#/gist/3fcc8d04411029140158775ce1a79823/097a692b04ffdf80c84344993789737fa9bb9ce6)). Note that the location information in the JS AST will have to be incremented, to reflect its offset within the HTML document.
> * Do the same for CSS

**Comment by joemaller** [comment]
This sounds like it would also work for Vue.js [single file components](https://vuejs.org/v2/guide/single-file-components.html) which contain JS, CSS and HTML-ish blocks.

**Comment by pomber** [comment]
How do you suggest detecting this (styled-components):  
```jsx
const Button = styled.button`
	color: palevioletred;
	font-size: 1em;
`;

// We're extending Button with some extra styles
const TomatoButton = Button.extend`
	color: tomato;
	border-color: tomato;
`;
```

or this (styled-jsx):
```jsx
const button = `button { color: hotpink; }`
const styles = `div { color: green; }`
export default () => (
  <div>
    <button>styled-jsx</button>
    <style jsx>{styles}</style>
    <style jsx>{button}</style>
  </div>
)
```

**Comment by vjeux** [comment]
For 1), we can hardcode `X.extend` hoping that it's not going to trigger on other files.

For 2), I don't think that we can support it, if people write code like this, too bad they won't be able to use prettier.

**Comment by vjeux** [comment]
@mxstbr for 1), have you considered an API that's more amenable to static analysis?

**Comment by mxstbr** [comment]
@vjeux I'm not aware of literally any other library doing uppercase variable `.extend` with a tagged template literal. Quite honestly, given the "popularity" of tagged template literals outside of GraphQL and styled-components (nobody uses them hah) I doubt anybody's going to use that :wink:

That being said, we're open to suggestions that preserve the DX but make it easier to statically analyse, /cc @geelen

**Comment by giuseppeg** [comment]
@pomber 2) is not supported in styled-jsx, nor are dynamic styles yet eg `${(props) => props.something}`. We do support external styles:

```jsx
export default `
  div { color: red }
`
```

What we do is that whenever we encounter a `TemplateLiteral` export we try to parse it with csstree (much faster than postcss), if parsing succeed then we treat it as styles – it is not the best solution but makes it possible to support the best api.

**Comment by philpl** [comment]
I think for 1) to not have any side effects you'll need to keep a whitelist of imports that switch the css tagged template literal on. For 2) there's probably only the all or nothing way unfortunately.

You could also build in a heuristic for tagged template literals that look like they contain css

**Comment by vjeux** [comment]
Done in 1.5
